### PR TITLE
Uses raytrace to clamp drop distance

### PR DESCRIPTION
**Pull request by Tomeno**
also fixes interaction range bug thing

### CODE REVIEW

**Comment by ComicIronic** [comment]
What didn't work in the existing code?

**Comment by Tomeno** [comment]
the logic was wrong. if you attempted to drop beyond your range, you would drop the item at your location instead. after these changes, you drop the item at the farthest possible point in the direction of the target.

File: Content.Shared/GameObjects/EntitySystems/SharedInteractionSystem.cs [hunk 1]
```diff
@@ -26,6 +26,43 @@ public class SharedInteractionSystem : EntitySystem
...
+        {
+            var dir = otherCoords.Position - coords.Position;
+
+            if (dir.LengthSquared.Equals(0f)) return 0f;
+
+            var ray = new CollisionRay(coords.Position, dir.Normalized, collisionMask);
+            var rayResults = _physicsManager.IntersectRayWithPredicate(coords.MapId, ray, dir.Length, predicate, returnOnFirstHit: false).ToList();
+
+            if (rayResults.Count == 0) return dir.Length;
+            return (rayResults[0].HitPos - coords.Position).Length;
```
[the remark this answers is not in the document]
Comment (Tomeno) [reply Content.Shared/GameObjects/EntitySystems/SharedInteractionSystem.cs:49]: it's not actually ordered by impact order, if there are multiple impacts on the trace then any one might be returned as "first" and the item ends up in a wall. same thing applies to interactions where they may be incorrectly considered as possible.

File: Content.Server/GameObjects/EntitySystems/HandsSystem.cs [hunk 2]
```diff
@@ -18,6 +18,7 @@
 using Robust.Shared.Localization;
 using Robust.Shared.Map;
 using Robust.Shared.Players;
+using System;
```
[the remark this answers is not in the document]
Comment (Tomeno) [reply Content.Server/GameObjects/EntitySystems/HandsSystem.cs:21]: yes.

[the remark this answers is not in the document]
Comment (PJB3005) [reply Content.Server/GameObjects/EntitySystems/HandsSystem.cs:21]: This PR adds a reference to `System.Math` so...

File: Content.Shared/GameObjects/EntitySystems/SharedInteractionSystem.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ComicIronic) [reply Content.Shared/GameObjects/EntitySystems/SharedInteractionSystem.cs:49]: @InquisitivePenguin This is precisely the problem that caused #1014 . "First hit" is only useful if all you care about is *some* collision. Getting the closest collision requires calculating all of them.

[the remark this answers is not in the document]
Comment (InquisitivePenguin) [reply Content.Shared/GameObjects/EntitySystems/SharedInteractionSystem.cs:49]: But my point is that they're not being sorted in any way or being calculated. It doesn't matter anyway, this PR is already merged.

[the remark this answers is not in the document]
Comment (ComicIronic) [reply Content.Shared/GameObjects/EntitySystems/SharedInteractionSystem.cs:49]: The raycast results are sorted at the end of `IntersectRayWithPredicate`, in order of distance from the ray source. This is the correct way to obtain the closest ray collision.

[the remark this answers is not in the document]
Comment (InquisitivePenguin) [reply Content.Shared/GameObjects/EntitySystems/SharedInteractionSystem.cs:49]: Oh you're right, I forgot about that. Never mind then.

**Pull request merged**